### PR TITLE
[ws client]: Make requests and subscriptions cancel-safe

### LINKED ISSUE #233: [ws client]: don't cancel background thread in case a request has been dropped
It would seem that failing to send a response on request terminates the client: https://github.com/paritytech/jsonrpsee/blob/cf2a5a7750baee0e1bc7693ec24e482fc162327d/ws-client/src/client.rs#L381

Can this be handled? Probably we don't want to terminate the client if a single connection is dropped.

Let me know if I missed anything!

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

Closes #233

### CODE REVIEW

**Review by niklasad1 — APPROVED** [review]
Thanks, I think we need don't need the same for subscriptions [because of this](https://github.com/paritytech/jsonrpsee/pull/237/files#diff-d54065c50060788f57a654fb34858be842053136d4d0883a91ff5b2af25152faR404-R408)

A bonus would be to have tests for this too :)

**Comment by niklasad1** [comment]
I create a PR to your branch that simplifies the code and takes subscriptions into consideration too.

**Review by niklasad1 — COMMENTED** [review]
We need to fix this for subscriptions too,

I created a PR to your branch that simplifies the code and takes subscriptions into consideration too.

**Review by niklasad1 — APPROVED** [review]

File: ws-client/src/client.rs [hunk 1]
```diff
@@ -373,38 +373,33 @@ fn process_response(
...
-			let send_back_oneshot = manager.complete_pending_call(response_id).ok_or(Error::InvalidRequestId)?;
+			let send_back_oneshot = match manager.complete_pending_call(response_id) {
+				Some(Some(send)) => send,
+				Some(None) => return Ok(None),
+				None => return Err(Error::InvalidRequestId),
+			};
+
 			manager.reclaim_request_id(response_id);
 			let response = response.try_into().map_err(Error::Request);
-			match send_back_oneshot.map(|tx| tx.send(response)) {
```
Comment (niklasad1) [inline ws-client/src/client.rs:379]: This was very stupid indeed by me, basically if it's fails it returns the message that was sent.

So fine to just ignore this.

**Comment by gregdhill** [comment]
@niklasad1 can we merge this?

**Comment by niklasad1** [comment]
> @niklasad1 can we merge this? 

It looks good to me but we require two approvals for pull requests.

**Review by dvdplm — APPROVED** [review]

**Pull request merged**